File: src/bmsx_cpp/render/3d/mesh/vertex_stream.cpp
#include "render/3d/mesh/vertex_stream.h"

#include "machine/devices/vdp/fixed_point.h"

#include <vector>

namespace bmsx {
namespace {

struct ResolvedMeshMaterial {
	std::array<f32, 4> color{1.0f, 1.0f, 1.0f, 1.0f};
	MeshDrawMaterial draw{};
};

enum class SkinVectorKind {
	Position,
	Normal,
};

void accumulateWeightedVector(Vec3& weighted, const Vec3& transformed, f32 weight) {
	weighted.x += transformed.x * weight;
	weighted.y += transformed.y * weight;
	weighted.z += transformed.z * weight;
}

std::array<f32, 4> packetColorAsLinear(u32 color) {
	return {
		static_cast<f32>((color >> 16u) & 0xffu) / 255.0f,
		static_cast<f32>((color >> 8u) & 0xffu) / 255.0f,
		static_cast<f32>(color & 0xffu) / 255.0f,
		static_cast<f32>((color >> 24u) & 0xffu) / 255.0f,
	};
}

size_t meshMorphTargetCount(const VdpMeshSourceMesh& mesh, const GameView::VdpMeshRenderEntry& entry) {
	size_t morphCount = entry.morphCount;
	if (morphCount > mesh.morphPositions->size()) {
		morphCount = mesh.morphPositions->size();
	}
	return morphCount;
}

bool meshHasSkinningSource(const VdpMeshSourceMesh& mesh, const GameView::VdpMeshRenderEntry& entry) {
	const size_t influenceCount = (mesh.positions->size() / 3u) * 4u;
	return entry.jointCount != 0u && mesh.jointIndices->size() >= influenceCount && mesh.jointWeights->size() >= influenceCount;
}

template<SkinVectorKind Kind>
void writeSkinnedVectorInto(Vec3& out,
							const std::array<Render3D::Mat4, VDP_JTU_MATRIX_COUNT>& jointMatrices,
							const VdpMeshSourceMesh& mesh,
							size_t vertexIndex,
							f32 x,
							f32 y,
							f32 z,
							u32 jointCount) {
	const size_t influenceBase = vertexIndex * 4u;
	const std::vector<u16>& jointIndices = *mesh.jointIndices;
	const std::vector<f32>& jointWeights = *mesh.jointWeights;
	Vec3 weighted{};
	Vec3 transformed{};
	for (size_t influence = 0u; influence < 4u; ++influence) {
		const u16 joint = jointIndices[influenceBase + influence];
		if (joint < jointCount) {
			if constexpr (Kind == SkinVectorKind::Position) {
				Render3D::mat4TransformAffinePoint3Into(transformed, jointMatrices[joint], x, y, z);
			} else {
				Render3D::mat4TransformDir3Into(transformed, jointMatrices[joint], x, y, z);
			}
		} else {
			Render3D::vec3Set(transformed, x, y, z);
		}
		accumulateWeightedVector(weighted, transformed, jointWeights[influenceBase + influence]);
	}
	out = weighted;
}

void writeMeshVertex(MeshStreamVertex& target,
					 const std::array<Render3D::Mat4, VDP_JTU_MATRIX_COUNT>& jointMatrices,
					 const std::array<f32, VDP_MDU_MORPH_WEIGHT_LIMIT>& morphWeights,
					 const VdpMeshSourceMesh& mesh,
					 size_t vertexIndex,
					 size_t morphCount,
					 bool skinningEnabled,
					 u32 jointCount,
					 const ResolvedMeshMaterial& material) {
	const size_t positionBase = vertexIndex * 3u;
	const std::vector<f32>& positions = *mesh.positions;
	const std::vector<f32>& normals = *mesh.normals;
	const std::vector<std::vector<f32>>& morphPositions = *mesh.morphPositions;
	const std::vector<std::vector<f32>>& morphNormals = *mesh.morphNormals;
	f32 x = positions[positionBase];
	f32 y = positions[positionBase + 1u];
	f32 z = positions[positionBase + 2u];
	f32 nx = 0.0f;
	f32 ny = 0.0f;
	f32 nz = 1.0f;
	if (positionBase + 2u < normals.size()) {
		nx = normals[positionBase];
		ny = normals[positionBase + 1u];
		nz = normals[positionBase + 2u];
	}
	for (u32 morphIndex = 0u; morphIndex < morphCount; ++morphIndex) {
		const std::vector<f32>& morph = morphPositions[morphIndex];
		const f32 weight = morphWeights[morphIndex];
		x += morph[positionBase] * weight;
		y += morph[positionBase + 1u] * weight;
		z += morph[positionBase + 2u] * weight;
		if (morphIndex < morphNormals.size()) {
			const std::vector<f32>& morphNormal = morphNormals[morphIndex];
			nx += morphNormal[positionBase] * weight;
			ny += morphNormal[positionBase + 1u] * weight;
			nz += morphNormal[positionBase + 2u] * weight;
		}
	}
	if (skinningEnabled) {
		Vec3 skinnedPosition{};
		writeSkinnedVectorInto<SkinVectorKind::Position>(skinnedPosition, jointMatrices, mesh, vertexIndex, x, y, z, jointCount);
		x = skinnedPosition.x;
		y = skinnedPosition.y;
		z = skinnedPosition.z;
		Vec3 skinnedNormal{};
		writeSkinnedVectorInto<SkinVectorKind::Normal>(skinnedNormal, jointMatrices, mesh, vertexIndex, nx, ny, nz, jointCount);
		nx = skinnedNormal.x;
		ny = skinnedNormal.y;
		nz = skinnedNormal.z;
	}
	const size_t uvBase = vertexIndex * 2u;
	const std::vector<f32>& texcoords = *mesh.texcoords;
	const bool hasTexcoord = uvBase + 1u < texcoords.size();
	const size_t colorBase = vertexIndex * 4u;
	const std::vector<f32>& colors = *mesh.colors;
	const bool hasColor = colorBase + 3u < colors.size();
	target.x = x;
	target.y = y;
	target.z = z;
	target.nx = nx;
	target.ny = ny;
	target.nz = nz;
	target.u = hasTexcoord ? texcoords[uvBase] : 0.0f;
	target.v = hasTexcoord ? texcoords[uvBase + 1u] : 0.0f;
	target.r = hasColor ? colors[colorBase] * material.color[0] : material.color[0];
	target.g = hasColor ? colors[colorBase + 1u] * material.color[1] : material.color[1];
	target.b = hasColor ? colors[colorBase + 2u] * material.color[2] : material.color[2];
	target.a = hasColor ? colors[colorBase + 3u] * material.color[3] : material.color[3];
}

i32 meshSurfaceMode(VdpMeshAlphaMode alphaMode) {
	switch (alphaMode) {
		case VdpMeshAlphaMode::Opaque: return MESH_SURFACE_OPAQUE;
		case VdpMeshAlphaMode::Mask: return MESH_SURFACE_MASK;
		case VdpMeshAlphaMode::Blend: return MESH_SURFACE_BLEND;
	}
	return MESH_SURFACE_OPAQUE;
}

ResolvedMeshMaterial resolveMeshMaterial(const VdpMeshSourceMaterial& sourceMaterial, const GameView::VdpMeshRenderEntry& entry) {
	ResolvedMeshMaterial material;
	material.color = packetColorAsLinear(entry.color);
	material.color[0] *= sourceMaterial.baseColorFactor[0];
	material.color[1] *= sourceMaterial.baseColorFactor[1];
	material.color[2] *= sourceMaterial.baseColorFactor[2];
	material.color[3] *= sourceMaterial.baseColorFactor[3];
	material.draw.surface = meshSurfaceMode(sourceMaterial.alphaMode);
	material.draw.alphaCutoff = sourceMaterial.alphaCutoff;
	material.draw.metallicFactor = sourceMaterial.metallicFactor;
	material.draw.roughnessFactor = sourceMaterial.roughnessFactor;
	material.draw.emissiveFactor = sourceMaterial.emissiveFactor;
	material.draw.doubleSided = sourceMaterial.doubleSided;
	material.draw.unlit = sourceMaterial.unlit;
	return material;
}

} // namespace

MeshDrawStream MeshVertexStreamBuilder::build(const GameView& view,
											 const VdpMeshSourceMesh& mesh,
											 const VdpMeshSourceMaterial& sourceMaterial,
											 const GameView::VdpMeshRenderEntry& entry) {
	decodeSignedQ16_16WordsInto(m_modelMatrix.data(), view.vdpXfMatrixWords.data() + static_cast<size_t>(entry.modelMatrixIndex * VDP_XF_MATRIX_WORDS), VDP_XF_MATRIX_WORDS);
	Render3D::mat4Normal3Into(m_normalMatrix, m_modelMatrix);
	const ResolvedMeshMaterial material = resolveMeshMaterial(sourceMaterial, entry);
	const size_t sourceVertexCount = mesh.hasIndices ? mesh.indices->size() : mesh.positions->size() / 3u;
	const size_t outputVertexCount = sourceVertexCount > VDP_MDU_VERTEX_LIMIT ? VDP_MDU_VERTEX_LIMIT : sourceVertexCount;
	const size_t morphCount = meshMorphTargetCount(mesh, entry);
	decodeSignedQ16_16WordsInto(m_morphWeights.data(), view.vdpMorphWeightWords.data() + entry.morphBase, morphCount);
	const bool skinningEnabled = meshHasSkinningSource(mesh, entry);
	if (skinningEnabled) {
		for (u32 index = 0u; index < entry.jointCount; ++index) {
			decodeSignedQ16_16WordsInto(m_jointMatrices[index].data(), view.vdpJointMatrixWords.data() + static_cast<size_t>((entry.jointBase + index) * VDP_JTU_MATRIX_WORDS), VDP_JTU_MATRIX_WORDS);
		}
	}
	m_vertexCount = outputVertexCount;
	if (!mesh.hasIndices) {
		for (size_t index = 0u; index < outputVertexCount; ++index) {
			writeMeshVertex(m_vertices[index], m_jointMatrices, m_morphWeights, mesh, index, morphCount, skinningEnabled, entry.jointCount, material);
		}
	} else {
		const std::vector<u32>& indices = *mesh.indices;
		for (size_t index = 0u; index < outputVertexCount; ++index) {
			writeMeshVertex(m_vertices[index], m_jointMatrices, m_morphWeights, mesh, indices[index], morphCount, skinningEnabled, entry.jointCount, material);
		}
	}
	return {m_vertices.data(), m_vertexCount, &m_modelMatrix, &m_normalMatrix, material.draw};
}

} // namespace bmsx

File: src/bmsx/render/3d/mesh/vertex_stream.ts
import {
	VDP_JTU_MATRIX_COUNT,
	VDP_JTU_MATRIX_WORDS,
	VDP_MDU_MORPH_WEIGHT_LIMIT,
	VDP_MDU_VERTEX_LIMIT,
} from '../../../machine/devices/vdp/contracts';
import { decodeSignedQ16_16WordsInto } from '../../../machine/devices/vdp/fixed_point';
import {
	VDP_MESH_ALPHA_BLEND,
	VDP_MESH_ALPHA_MASK,
	type VdpMeshSourceMaterial,
	type VdpMeshSourceMesh,
} from '../../../machine/devices/vdp/mesh_source';
import { VDP_XF_MATRIX_WORDS } from '../../../machine/devices/vdp/xf';
import type { GameView } from '../../gameview';
import { M4 } from '../math';

export const MESH_SURFACE_OPAQUE = 0;
export const MESH_SURFACE_MASK = 1;
export const MESH_SURFACE_BLEND = 2;
export const MESH_VERTEX_FLOATS = 12;
export const MESH_POSITION_FLOAT_OFFSET = 0;
export const MESH_NORMAL_FLOAT_OFFSET = 3;
export const MESH_UV_FLOAT_OFFSET = 6;
export const MESH_COLOR_FLOAT_OFFSET = 8;

export interface ResolvedMeshMaterial {
	color0: number;
	color1: number;
	color2: number;
	color3: number;
	surface: number;
	alphaCutoff: number;
	metallicFactor: number;
	roughnessFactor: number;
	emissive0: number;
	emissive1: number;
	emissive2: number;
	doubleSided: boolean;
	unlit: boolean;
}

export class MeshVertexStreamBuilder {
	readonly vertices = new Float32Array(VDP_MDU_VERTEX_LIMIT * MESH_VERTEX_FLOATS);
	readonly modelMatrix = new Float32Array(16);
	readonly normalMatrix = new Float32Array(9);
	readonly material: ResolvedMeshMaterial = {
		color0: 1,
		color1: 1,
		color2: 1,
		color3: 1,
		surface: MESH_SURFACE_OPAQUE,
		alphaCutoff: 0.5,
		metallicFactor: 1,
		roughnessFactor: 1,
		emissive0: 0,
		emissive1: 0,
		emissive2: 0,
		doubleSided: false,
		unlit: false,
	};
	vertexCount = 0;

	private readonly jointMatrices = new Float32Array(VDP_JTU_MATRIX_COUNT * VDP_JTU_MATRIX_WORDS);
	private readonly morphWeights = new Float32Array(VDP_MDU_MORPH_WEIGHT_LIMIT);

	build(view: GameView, mesh: VdpMeshSourceMesh, sourceMaterial: VdpMeshSourceMaterial, entryIndex: number): void {
		const modelMatrixIndex = view.vdpMeshModelMatrixIndex[entryIndex];
		decodeSignedQ16_16WordsInto(this.modelMatrix, 0, view.vdpXfMatrixWords, modelMatrixIndex * VDP_XF_MATRIX_WORDS, VDP_XF_MATRIX_WORDS);
		M4.normal3Into(this.normalMatrix, this.modelMatrix);
		this.resolveMeshMaterial(sourceMaterial, view.vdpMeshColor[entryIndex]);
		const sourceVertexCount = mesh.hasIndices ? mesh.indices.length : mesh.positions.length / 3;
		const vertexCount = sourceVertexCount > VDP_MDU_VERTEX_LIMIT ? VDP_MDU_VERTEX_LIMIT : sourceVertexCount;
		const morphCount = this.meshMorphTargetCount(mesh, view.vdpMeshMorphCount[entryIndex]);
		decodeSignedQ16_16WordsInto(this.morphWeights, 0, view.vdpMorphWeightWords, view.vdpMeshMorphBase[entryIndex], morphCount);
		const jointCount = view.vdpMeshJointCount[entryIndex];
		const skinningEnabled = this.meshHasSkinningSource(mesh, jointCount);
		if (skinningEnabled) {
			const jointBase = view.vdpMeshJointBase[entryIndex];
			const words = view.vdpJointMatrixWords;
			for (let jointIndex = 0; jointIndex < jointCount; jointIndex += 1) {
				decodeSignedQ16_16WordsInto(this.jointMatrices, jointIndex * VDP_JTU_MATRIX_WORDS, words, (jointBase + jointIndex) * VDP_JTU_MATRIX_WORDS, VDP_JTU_MATRIX_WORDS);
			}
		}
		this.vertexCount = vertexCount;
		if (mesh.hasIndices) {
			const indices = mesh.indices;
			for (let index = 0; index < vertexCount; index += 1) {
				this.writeMeshVertex(mesh, indices[index], index * MESH_VERTEX_FLOATS, morphCount, skinningEnabled, jointCount);
			}
		} else {
			for (let index = 0; index < vertexCount; index += 1) {
				this.writeMeshVertex(mesh, index, index * MESH_VERTEX_FLOATS, morphCount, skinningEnabled, jointCount);
			}
		}
	}

	private meshSurfaceMode(alphaMode: number): number {
		switch (alphaMode) {
			case VDP_MESH_ALPHA_MASK: return MESH_SURFACE_MASK;
			case VDP_MESH_ALPHA_BLEND: return MESH_SURFACE_BLEND;
		}
		return MESH_SURFACE_OPAQUE;
	}

	private writePacketColor(color: number): void {
		const target = this.material;
		target.color0 = ((color >>> 16) & 0xff) / 255;
		target.color1 = ((color >>> 8) & 0xff) / 255;
		target.color2 = (color & 0xff) / 255;
		target.color3 = ((color >>> 24) & 0xff) / 255;
	}

	private resolveMeshMaterial(source: VdpMeshSourceMaterial, colorWord: number): void {
		const target = this.material;
		this.writePacketColor(colorWord);
		target.color0 *= source.baseColor0;
		target.color1 *= source.baseColor1;
		target.color2 *= source.baseColor2;
		target.color3 *= source.baseColor3;
		target.surface = this.meshSurfaceMode(source.alphaMode);
		target.alphaCutoff = source.alphaCutoff;
		target.metallicFactor = source.metallicFactor;
		target.roughnessFactor = source.roughnessFactor;
		target.emissive0 = source.emissive0;
		target.emissive1 = source.emissive1;
		target.emissive2 = source.emissive2;
		target.doubleSided = source.doubleSided;
		target.unlit = source.unlit;
	}

	private writeMeshVertex(mesh: VdpMeshSourceMesh,
		vertexIndex: number,
		outputBase: number,
		morphCount: number,
		skinningEnabled: boolean,
		jointCount: number): void {
		const positionBase = vertexIndex * 3;
		let x = mesh.positions[positionBase];
		let y = mesh.positions[positionBase + 1];
		let z = mesh.positions[positionBase + 2];
		let nx = 0;
		let ny = 0;
		let nz = 1;
		const normals = mesh.normals;
		if (positionBase + 2 < normals.length) {
			nx = normals[positionBase];
			ny = normals[positionBase + 1];
			nz = normals[positionBase + 2];
		}
		const morphPositions = mesh.morphPositions;
		const morphNormals = mesh.morphNormals;
		for (let morphIndex = 0; morphIndex < morphCount; morphIndex += 1) {
			const morph = morphPositions[morphIndex];
			const weight = this.morphWeights[morphIndex];
			x += morph[positionBase] * weight;
			y += morph[positionBase + 1] * weight;
			z += morph[positionBase + 2] * weight;
			if (morphIndex < morphNormals.length) {
				const morphNormal = morphNormals[morphIndex];
				nx += morphNormal[positionBase] * weight;
				ny += morphNormal[positionBase + 1] * weight;
				nz += morphNormal[positionBase + 2] * weight;
			}
		}
		if (skinningEnabled) {
			let weightedX = 0;
			let weightedY = 0;
			let weightedZ = 0;
			let weightedNx = 0;
			let weightedNy = 0;
			let weightedNz = 0;
			const influenceBase = vertexIndex * 4;
			const jointIndices = mesh.jointIndices;
			const jointWeights = mesh.jointWeights;
			for (let influence = 0; influence < 4; influence += 1) {
				const joint = jointIndices[influenceBase + influence];
				const weight = jointWeights[influenceBase + influence];
				const scratchBase = outputBase;
				if (joint < jointCount) {
					const matrixBase = joint * VDP_JTU_MATRIX_WORDS;
					M4.transformAffinePoint3At(this.vertices, scratchBase, this.jointMatrices, matrixBase, x, y, z);
					M4.transformDir3At(this.vertices, scratchBase + 3, this.jointMatrices, matrixBase, nx, ny, nz);
				} else {
					this.vertices[scratchBase] = x;
					this.vertices[scratchBase + 1] = y;
					this.vertices[scratchBase + 2] = z;
					this.vertices[scratchBase + 3] = nx;
					this.vertices[scratchBase + 4] = ny;
					this.vertices[scratchBase + 5] = nz;
				}
				weightedX += this.vertices[scratchBase] * weight;
				weightedY += this.vertices[scratchBase + 1] * weight;
				weightedZ += this.vertices[scratchBase + 2] * weight;
				weightedNx += this.vertices[scratchBase + 3] * weight;
				weightedNy += this.vertices[scratchBase + 4] * weight;
				weightedNz += this.vertices[scratchBase + 5] * weight;
			}
			x = weightedX;
			y = weightedY;
			z = weightedZ;
			nx = weightedNx;
			ny = weightedNy;
			nz = weightedNz;
		}
		const uvBase = vertexIndex * 2;
		const texcoords = mesh.texcoords;
		const hasTexcoord = uvBase + 1 < texcoords.length;
		const colorBase = vertexIndex * 4;
		const colors = mesh.colors;
		const hasColor = colorBase + 3 < colors.length;
		const vertices = this.vertices;
		const material = this.material;
		vertices[outputBase] = x;
		vertices[outputBase + 1] = y;
		vertices[outputBase + 2] = z;
		vertices[outputBase + 3] = nx;
		vertices[outputBase + 4] = ny;
		vertices[outputBase + 5] = nz;
		vertices[outputBase + 6] = hasTexcoord ? texcoords[uvBase] : 0;
		vertices[outputBase + 7] = hasTexcoord ? texcoords[uvBase + 1] : 0;
		vertices[outputBase + 8] = hasColor ? colors[colorBase] * material.color0 : material.color0;
		vertices[outputBase + 9] = hasColor ? colors[colorBase + 1] * material.color1 : material.color1;
		vertices[outputBase + 10] = hasColor ? colors[colorBase + 2] * material.color2 : material.color2;
		vertices[outputBase + 11] = hasColor ? colors[colorBase + 3] * material.color3 : material.color3;
	}

	private meshMorphTargetCount(mesh: VdpMeshSourceMesh, packetMorphCount: number): number {
		const morphPositions = mesh.morphPositions;
		return packetMorphCount < morphPositions.length ? packetMorphCount : morphPositions.length;
	}

	private meshHasSkinningSource(mesh: VdpMeshSourceMesh, jointCount: number): boolean {
		const influenceCount = (mesh.positions.length / 3) * 4;
		return jointCount !== 0 && mesh.jointIndices.length >= influenceCount && mesh.jointWeights.length >= influenceCount;
	}
}
